### DISCUSSION TITLE
[Done] Replace Connect with Express?

### DISCUSSION THREAD

**Discussion by StorytellerCZ** [meteor/meteor#12410] [Feature Requests]
Under the hood Meteor uses [Connect](https://github.com/senchalabs/connect). But Connect hasn't seen any development in years and any new server side options focus on Express. So I would like to start a discussion on this topic if we maybe shouldn't at some point migrate to use Express? This because the API of these two is slowly diverging so it won't be possible to use Express middleware and so on with Meteor in the future.

Or maybe there are other things that we can do on this front. This is has come as I'm in the middle of upgrading to Apollo server v4.

**Comment by radekmie** [comment]
If needed, you can always use Express as a middleware for Connect - that's what I did when migrating to Apollo Server v4. Overall, I think Express would be a nice upgrade, but I'm not sure if it's necessary.

**Reply by StorytellerCZ** [reply]
Can you update the Apollo skeleton with your solution?

**Reply by radekmie** [reply]
I won't have time for that soon, but it's as simple as this:
```ts
// Expose the server through Express due to API compatibility.
WebApp.connectHandlers.use(
  '/graphql',                                     // Configure the path as you want.
  express()                                       // Create new Express router.
    .disable('etag')                              // We don't server GET requests, so there's no need for that.
    .disable('x-powered-by')                      // A small safety measure.
    .use(json())                                  // From `body-parser`.
    .use(expressMiddleware(server, { context })), // From `@apollo/server/express4`.
);
```

**Reply by StorytellerCZ** [reply]
Thank you! I'll take it from here. Still an issue that we have to establish Express on top of Connect though.

**Comment by Grubba27** [comment]
I had similar thoughts these days. 
I think it would be nice to have something like express or fastify under the hood but something like h3 that is the engine that the they use on nuxt in their framework sounds and feels nice.


At the same time having the possibility of fully using the express or fastify ecosystem sounds a good selling point as well. I think guys at NestJS let you choose which of them you want as engine and I think that their DX is awesome


I do not know fully how much we use connect at Meteor and what other gains we could earn by switching and how hard would it be

**Reply by radekmie** [reply]
> I do not know fully how much we use connect at Meteor and what other gains we could earn by switching and how hard would it be
 
There's not much in Core packages, but it'd affect most apps that expose some APIs - REST, GraphQL, webhook handlers - either directly (in their code) or indirectly (packages like `communitypackages:picker`). I don't think it'd be a lot of work for them to update, though, as the middlewares are (mostly) compatible.

The main gain would be compatibility with modern middlewares. Performance-wise, most of the Meteor traffic goes through DDP, so it won't help that much.

**Reply by StorytellerCZ** [reply]
If I remember correctly including Express would make Picker unnecessary.

**Reply by radekmie** [reply]
Yes and no - we're not using Picker anywhere and yet expose all sort of HTTP endpoints. It'd be easier with Express or Fastify, though.

**Reply by henriquealbert** [reply]
I don't know if it's worth the time working on it, but it would be nice to see what gains we would have if we switched to Express/Fastify/H3

**Reply by radekmie** [reply]
I don't think so either - it's something that can be changed later. It'll be a breaking change too, though.

**Reply by StorytellerCZ** [reply]
I didn't meant that this should be done in the near future. It is rather something to look for towards the end of 2023 (at earliest) and beyond.

**Comment by rj-david** [comment]
Seems this will happen soon

https://github.com/meteor/meteor/pull/12442

@Grubba27, I am unsure how big of a change this will be on those like us with a number of middlewares. Is there a way to do this before Meteor 3.0 and not together with 3.0?

Just a way to spread out the needed migrations.

(From the PR text, seems like post Meteor 3.0 is outside the question)

**Reply by Grubba27** [reply]
Hello @rj-david!
here are my toughts:

> I am unsure how big of a change this will be on those like us with a number of middlewares. Is there a way to do this before Meteor 3.0 and not together with 3.0

It is going together with 3.0 because it is a kind of breaking change, and it is time we have to make a breaking change like this one.
For the middleware support, from what I can tell, their API is **_extremely similar_**, all the tests I've run in Meteor Codebase passed, and for most cases, will be okay.

**Reply by rj-david** [reply]
Time for us to unify the implementation of our webhooks. Hopefully, most of the modules we used have corresponding modules in express. And maybe time to move away from restivus for the very old apis

This might also help in the documentation: https://expressjs.com/en/guide/migrating-4.html
